File: GolangApi/controller/user_controller.go
package controller

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"GolangApi/database"
	"GolangApi/models"
)

// GetUsers: lista usuarios
func GetUsers(c *gin.Context) {
	query := `SELECT id, name, email, username, password, created_at, updated_at FROM users ORDER BY id`
	rows, err := database.DB.Query(query)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao buscar usuários"})
		return
	}
	defer rows.Close()

	var users []models.User
	for rows.Next() {
		var user models.User
		err := rows.Scan(
			&user.ID,
			&user.Name,
			&user.Email,
			&user.Username,
			&user.Password,
			&user.CreatedAt,
			&user.UpdatedAt,
		)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao processar usuários"})
			return
		}
		users = append(users, user)
	}

	c.JSON(http.StatusOK, users)
}

// GetUserByID - busca usuario por id
func GetUserByID(c *gin.Context) {
	id := c.Param("id")
	query := `SELECT id, name, email, username, password, created_at, updated_at FROM users WHERE id = $1`

	var user models.User
	err := database.DB.QueryRow(query, id).Scan(
		&user.ID,
		&user.Name,
		&user.Email,
		&user.Username,
		&user.Password,
		&user.CreatedAt,
		&user.UpdatedAt,
	)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "Usuário não encontrado"})
		return
	}

	c.JSON(http.StatusOK, user)
}


// CreateUser - cria um usuário
func CreateUser(c *gin.Context) {
	var user models.User
	if err := c.ShouldBindJSON(&user); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	query := `INSERT INTO users (name, email, username, password) VALUES ($1, $2, $3, $4) RETURNING id`
	err := database.DB.QueryRow(query, user.Name, user.Email, user.Username, user.Password).Scan(&user.ID)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao criar usuário"})
		return
	}

	c.JSON(http.StatusCreated, user)
}

// UpdateUser - atualiza um usuario
func UpdateUser(c *gin.Context) {
	id := c.Param("id")
	var user models.User
	if err := c.ShouldBindJSON(&user); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	query := `UPDATE users SET name = $1, email = $2, username = $3 WHERE id = $4`
	result, err := database.DB.Exec(query, user.Name, user.Email, user.Username, id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao atualizar usuário"})
		return
	}

	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Usuário não encontrado"})
		return
	}

	userID, _ := strconv.Atoi(id)
	user.ID = userID
	c.JSON(http.StatusOK, user)
}

// DeleteUser - deleta um usuário
func DeleteUser(c *gin.Context) {
	id := c.Param("id")
	query := `DELETE FROM users WHERE id = $1`

	result, err := database.DB.Exec(query, id)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Erro ao deletar usuário"})
		return
	}

	rowsAffected, _ := result.RowsAffected()
	if rowsAffected == 0 {
		c.JSON(http.StatusNotFound, gin.H{"error": "Usuário não encontrado"})
		return
	}

	c.JSON(http.StatusNoContent, nil)
}

File: NodeApi/src/controllers/userController.ts
import { Request, Response } from 'express';
import User from '../models/User';

export class UserController {
  // POST /users - cria usuário
  async create(req: Request, res: Response): Promise<void> {
    try {
      const { name, email, username, password } = req.body;

      if (!name || !email || !username || !password) {
        res.status(400).json({ 
          error: 'Todos os campos são obrigatórios: name, email, username, password' 
        });
        return;
      }

      const newUser = await User.create({ name, email, username, password });
      res.status(201).json(newUser);
    } catch (error: any) {
      if (error.name === 'SequelizeUniqueConstraintError') {
        res.status(409).json({ error: 'Email ou username já cadastrado' });
        return;
      }
      console.error('Erro ao criar usuário:', error);
      res.status(500).json({ error: 'Erro ao criar usuário' });
    }
  }

  // GET /users - busca todos os usuários
  async list(req: Request, res: Response): Promise<void> {
    try {
      const users = await User.findAll({
        order: [['id', 'ASC']],
      });
      res.status(200).json(users);
    } catch (error) {
      console.error('Erro ao listar usuários:', error);
      res.status(500).json({ error: 'Erro ao listar usuários' });
    }
  }

  // GET /users/:id - busca usuario por ID
  async getById(req: Request, res: Response): Promise<void> {
    try {
      const { id } = req.params;
      const user = await User.findByPk(id);

      if (!user) {
        res.status(404).json({ error: 'Usuário não encontrado' });
        return;
      }

      res.status(200).json(user);
    } catch (error) {
      console.error('Erro ao buscar usuário:', error);
      res.status(500).json({ error: 'Erro ao buscar usuário' });
    }
  }

  // PUT /users/:id - atualiza usuario
  async update(req: Request, res: Response): Promise<void> {
    try {
      const { id } = req.params;
      const { name, email, username, password } = req.body;

      const userToUpdate = await User.findByPk(id);

      if (!userToUpdate) {
        res.status(404).json({ error: 'Usuário não encontrado' });
        return;
      }

      await userToUpdate.update({
        name: name || userToUpdate.name,
        email: email || userToUpdate.email,
        username: username || userToUpdate.username,
        password: password || userToUpdate.password,
      });

      res.status(200).json(userToUpdate);
    } catch (error: any) {
      if (error.name === 'SequelizeUniqueConstraintError') {
        res.status(409).json({ error: 'Email ou username já cadastrado' });
        return;
      }
      console.error('Erro ao atualizar usuário:', error);
      res.status(500).json({ error: 'Erro ao atualizar usuário' });
    }
  }

  // DELETE /users/:id - deleta usuario baseado no ID
  async delete(req: Request, res: Response): Promise<void> {
    try {
      const { id } = req.params;
      const user = await User.findByPk(id);

      if (!user) {
        res.status(404).json({ error: 'Usuário não encontrado' });
        return;
      }

      await user.destroy();
      res.status(204).send();
    } catch (error) {
      console.error('Erro ao deletar usuário:', error);
      res.status(500).json({ error: 'Erro ao deletar usuário' });
    }
  }
}